### PR TITLE
Add AdamW `Optimizer` @open sesame 12/04 16:19

### PR DESCRIPTION
**Pull request by jangdan**
This commit implements the AdamW algorithm in the form of an `Optimizer`.

Some (potential) further tasks include:
- Merging the implementations of Adam and AdamW, so that `AdamW` inherits from `Adam`
- Adding more unit tests

### CODE REVIEW

**Comment by jijoongmoon** [comment]
This commit needs the commit body. Could you write details in the commit body?

**Comment by myungjoo** [comment]
Please add a short description about the commit in the commit message body. This helps when we need to browse git logs to debug.

In case you are not familiar with git, you can edit the commit message:
```
$ git commit --amend
```

**Comment by EunjuYang** [comment]
Thank you for your contribution 😀
NNTrainer follows a specific coding convention (i.e., specific clang format is provided), so you need to apply it to pass the CI. Please refer to [[this](https://github.com/nnstreamer/nntrainer/blob/main/docs/coding-convention.md)].

File: nntrainer/optimizers/adamw.cpp [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+#include <adamw.h>
+#include <nntrainer_error.h>
+#include <nntrainer_log.h>
+#include <node_exporter.h>
+#include <util_func.h>
+
+namespace nntrainer {
+
+AdamW::AdamW() : adam_props(PropsB1(), PropsB2(), PropsEpsilon(), TorchRef()) {
```
Comment (DonghakPark) [inline nntrainer/optimizers/adamw.cpp:26]: Adam case, when the torch_ref flag is provided, it handles separate operations to ensure that the results are as similar as possible to PyTorch.

I am curious whether Adamw also considered this case in adding torch_ref or simply did so to have the same properties as Adam. It seems like torch_ref has no meaning for AdamW.

```.cpp
  if (torch_ref) {
    Tensor denom = wv.apply<float>(sqrtFloat<float>);
    denom.divide_i(sqrtFloat(biasCorrection2));
    denom.add_i(epsilon);
    wm.divide(denom, x_grad);

    context.applyGradient(context.getLearningRate() / biasCorrection1);

  } else {
    std::function<double(double)> sqrtEps = [epsilon](double f) {
      return 1 / (sqrtDouble(f) + epsilon);
    };
```

File: test/ccapi/unittest_ccapi.cpp [hunk 2]
```diff
@@ -147,6 +147,7 @@ TEST(ccapi_optimizer, construct_01_n) {
 TEST(ccapi_optimizer, construct_02_p) {
   EXPECT_NO_THROW(ml::train::optimizer::Adam());
   EXPECT_NO_THROW(ml::train::optimizer::SGD());
+  EXPECT_NO_THROW(ml::train::optimizer::AdamW());
```
Comment (DonghakPark) [inline test/ccapi/unittest_ccapi.cpp:150]: Have you ever conducted any actual training tests other than ccapi construct test?
when i tested on MNIST Application Example i got  ```segmentation fault (core dumped)``` error

tested on this code
```.cpp
  std::unique_ptr<ml::train::Model> model;
  model = createModel(ml::train::ModelType::NEURAL_NET);
  model->load(config, ml::train::ModelFormat::MODEL_FORMAT_INI_WITH_BIN);
  model->compile();
  model->initialize();
  model->setDataset(ml::train::DatasetModeType::MODE_TRAIN, dataset_train);
  model->setDataset(ml::train::DatasetModeType::MODE_VALID, dataset_val);
  model->train();
```
with this ```ini``` file
```.ini
# Network Section : Network

[Model]
Type = NeuralNetwork                # Network Type : Regression, KNN, NeuralNetwork
Epochs = 1500                       # Epochs
Loss = cross                        # Loss function : mse (mean squared error)
                                    # cross ( for cross entropy )
# Save_Path = "mnist_model.bin"     # model path to save / read
batch_size = 32                     # batch size

[Optimizer]
Type = adamw
beta1 = 0.9                         # beta 1 for adam
beta2 = 0.999                       # beta 2 for adam
epsilon = 1e-7                      # epsilon for adam

# Layer Section : Name
[inputlayer]
Type = input
Input_Shape = 1:28:28

# Layer Section : Name
[conv2d_c1_layer]
Type = conv2d
input_layers = inputlayer
kernel_size = 5,5
bias_initializer=zeros
Activation=sigmoid
weight_initializer = xavier_uniform
filters = 6
stride = 1,1
padding = 0,0

[pooling2d_p1]
Type=pooling2d
input_layers = conv2d_c1_layer
pool_size = 2,2
stride =2,2
padding = 0,0
pooling = average

[conv2d_c2_layer]
Type = conv2d
input_layers = pooling2d_p1
kernel_size = 5,5
bias_initializer=zeros
Activation=sigmoid
weight_initializer = xavier_uniform
filters = 12
stride = 1,1
padding = 0,0

[pooling2d_p2]
Type=pooling2d
input_layers = conv2d_c2_layer
pool_size = 2,2
stride =2,2
padding = 0,0
pooling = average

[flatten]
Type=flatten
input_layers = pooling2d_p2

[outputlayer]
Type = fully_connected
input_layers=flatten
Unit = 10                           # Output Layer Dimension ( = Weight Width )
weight_initializer = xavier_uniform
bias_initializer = zeros
Activation = softmax                # activation : sigmoid, softmax
```

File: api/ccapi/include/optimizer.h [hunk 3]
```diff
@@ -35,6 +35,7 @@ class LearningRateScheduler;
 enum OptimizerType {
   ADAM = ML_TRAIN_OPTIMIZER_TYPE_ADAM,      /** adam */
   SGD = ML_TRAIN_OPTIMIZER_TYPE_SGD,        /** sgd */
+  ADAMW = ML_TRAIN_OPTIMIZER_TYPE_ADAMW,    /** AdamW */
   UNKNOWN = ML_TRAIN_OPTIMIZER_TYPE_UNKNOWN /** unknown */
 };
```
Comment (SeoHyungjun) [inline api/ccapi/include/optimizer.h:40]: Wouldn't it be better to write in alphabetical order unless there's a specific reason?

```suggestion
enum OptimizerType {
  ADAM = ML_TRAIN_OPTIMIZER_TYPE_ADAM,      /** adam */
  ADAMW = ML_TRAIN_OPTIMIZER_TYPE_ADAMW,    /** AdamW */
  SGD = ML_TRAIN_OPTIMIZER_TYPE_SGD,        /** sgd */
  UNKNOWN = ML_TRAIN_OPTIMIZER_TYPE_UNKNOWN /** unknown */
};
```

File: api/nntrainer-api-common.h [hunk 4]
```diff
@@ -88,6 +88,7 @@ typedef enum {
 typedef enum {
   ML_TRAIN_OPTIMIZER_TYPE_ADAM = 0, /**< Adam Optimizer */
   ML_TRAIN_OPTIMIZER_TYPE_SGD = 1, /**< Stochastic Gradient Descent Optimizer */
+  ML_TRAIN_OPTIMIZER_TYPE_ADAMW = 2, /**< AdamW Optimizer */
```
Comment (SeoHyungjun) [inline api/nntrainer-api-common.h:91]: Ditto

File: nntrainer/app_context.cpp [hunk 5]
```diff
@@ -29,6 +29,7 @@
 
 #include <adam.h>
 #include <sgd.h>
+#include <adamw.h>
```
Comment (SeoHyungjun) [inline nntrainer/app_context.cpp:32]: In our project, the option "SortIncludes" is enabled in the ".clang-format" file.
Therefore, the order of headers must be sorted unconditionally.

```suggestion
#include <adam.h>
#include <adamw.h>
#include <sgd.h>
```

**Review by SeoHyungjun — COMMENTED** [review]
Thank you for your contribution!

Please check out the comments below. :art:

**Review by baek2sm — CHANGES_REQUESTED** [review]
As others have pointed out, you need adding a commit body and making modifications to follow the clang format. Then you can pass the CI.

File: nntrainer/optimizers/adamw.cpp (hunk 1 above)
Comment (jangdan) [reply nntrainer/optimizers/adamw.cpp:26]: Indeed, the `"torch_ref"` property is not used (at all) in `AdamW`. The reasoning behind keeping it (and also retaining the identifier `adam_props`, to address @baek2sm's [comment](https://github.com/nnstreamer/nntrainer/pull/2780#discussion_r1825411607)) was that I thought it would be a very good idea to, in a later PR, have the `AdamW` class inherit from the `Adam` class, to reduce code duplication, as they share a lot of boilerplate code. Sorry for not making this clear.

Any thoughts on this? If this is not the way to go, I'd happily make the necessary changes.

FYI: PyTorch's [AdamW class](https://pytorch.org/docs/stable/_modules/torch/optim/adamw.html#AdamW) <ins>doesn't</ins> seem to inherit from its [Adam class](https://pytorch.org/docs/stable/_modules/torch/optim/adam.html#Adam).

File: api/ccapi/include/optimizer.h (hunk 3 above)
Comment (jangdan) [reply api/ccapi/include/optimizer.h:40]: Good idea. It wasn't particularly clear that that was the case, with only two entries.

File: api/nntrainer-api-common.h (hunk 4 above)
Comment (jangdan) [reply api/nntrainer-api-common.h:95]: Done.

Comment (jangdan) [reply api/nntrainer-api-common.h:95]: Done.

Comment (jangdan) [reply api/nntrainer-api-common.h:95]: Done.

File: nntrainer/app_context.cpp (hunk 5 above)
Comment (jangdan) [reply nntrainer/app_context.cpp]: Fixed, with clang-format.

File: test/ccapi/unittest_ccapi.cpp (hunk 2 above)
Comment (jangdan) [reply test/ccapi/unittest_ccapi.cpp:159]: The code should now actually work.

Thanks.

**Review by baek2sm — APPROVED** [review]

File: nntrainer/optimizers/adamw.cpp [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+  Optimizer::exportTo(exporter, method);
+}
+
+void AdamW::setProperty(const std::vector<std::string> &values) {
+  auto left = loadProperties(values, adam_props);
+  Optimizer::setProperty(left);
+}
+
+void AdamW::applyGradient(RunOptimizerContext &context) {
+  Tensor &x_grad = context.getGradient();
```
Comment (EunjuYang) [inline nntrainer/optimizers/adamw.cpp:55]: Code to process mixed precision is missing. Please refer to 
https://github.com/nnstreamer/nntrainer/blob/0d74457e18b97fa8d26702ec7663d6f6e9510ac2/nntrainer/optimizers/adam.cpp#L75-L84

**Review by EunjuYang — COMMENTED** [review]
Thanks for your contribution to implement `AdamW` optimizer.
`AdamW` is a varient of `Adam` that applies weight decay.
However, I couldn't find the part for this weight decay. Could you explain if I missed something? Please check the comments below. 
----------------
I missed baek2sm's comment above. I will leave the updated review again.

File: nntrainer/optimizers/adamw.cpp (hunk 6 above)
Comment (jijoongmoon) [reply nntrainer/optimizers/adamw.cpp:55]: yes. this needs to be considered. Please check it.

File: nntrainer/optimizers/adamw.cpp [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+  wv.divide_i(sqrtFloat(biasCorrection2));
+  std::function<double(double)> sqrtEps = [epsilon](double f) {
+    return 1 / (sqrtDouble(f) + epsilon);
+  };
+  Tensor &term = wv;
+  term.apply<float>(sqrtEps);
+  term.divide_i(biasCorrection1);
+  term.multiply_i(wm);
+  x_grad.add_i(term);
```
[the remark this answers is not in the document]
Comment (jangdan) [reply nntrainer/optimizers/adamw.cpp:84]: I think you are correct. Thanks for the catch!

File: nntrainer/optimizers/adamw.cpp (hunk 6 above)
Comment (jangdan) [reply nntrainer/optimizers/adamw.cpp]: This, too, has been addressed.

**Review by jijoongmoon — APPROVED** [review]
LGTM
